lib/medtronic600/medtronic600Simulator.js

```diff
@@ -73,291 +111,123 @@ class Medtronic600Simulator {
-                const duration = Date.parse(event.time) - Date.parse(this.currBasal.time);
-                this.currBasal.with_duration(duration);
-
-                resumeBasal.duration = this.currBasal.expectedDuration - this.currBasal.duration -
-                  this.currStatus.duration;
-              }
-              common.truncateDuration(this.currBasal, 'medtronic');
-              this.events.push(this.currBasal.done());
-
-              // check if suspended basal is crossing schedule changes
-              this.currBasal = event;
-              if (this.checkForScheduleChanges()) {
-                // we should remember to update resume basal's suppressed too
-                resumeBasal.suppressed = this.currBasal.suppressed.suppressed;
-              }
-              common.truncateDuration(this.currBasal, 'medtronic');
-              this.events.push(this.currBasal.done());
 
-              this.currBasal = resumeBasal;
-              return;
-            }
+        if (event.time > this.currBasal.basalEndTime) {
+          // After a temp basal, the next basal segment start event doesn't happen for up to
+          // a minute after the temp basal complete event, leaving small basal data gaps every
+          // time a temp basal finishes.
+          // It's also feasible that we won't get a basal segment start event at all if the user
+          // has cancelled a temp basal, and followed it up immediately with another temp basal.
+
+          const newDeviceTime = new Date(Date.parse(this.currBasal.deviceTime).valueOf() +
+            this.currBasal.duration);
+
+          if (this.currBasal.suppressed.type !== event.type ||
+            this.currBasal.suppressed.deliveryType !== event.deliveryType ||
+            this.currBasal.suppressed.rate !== event.rate ||
+            this.currBasal.suppressed.scheduleName !== event.scheduleName) {
+            // If the new scheduled basal event doesn't have the same profile as the suppressed
+            // basal, tack the suppressed basal onto the end of the temp basal before the new
+            // scheduled basal starts.
+            this.events.push(this.currBasal.done());
+
+            const restoredBasal = this.config.builder.makeScheduledBasal()
+              .with_time(this.currBasal.basalEndTime)
+              .with_deviceTime(sundial.formatDeviceTime(newDeviceTime))
+              .with_rate(this.currBasal.suppressed.rate)
+              .with_scheduleName(this.currBasal.suppressed.scheduleName)
+              .with_duration(Date.parse(event.time) - Date.parse(this.currBasal.basalEndTime))
+              .with_conversionOffset(this.currBasal.conversionOffset)
+              .with_timezoneOffset(this.currBasal.timezoneOffset);
+
+            this.currBasal = restoredBasal;
+          } else {
+            // Otherwise, just fix the start time of the new basal event.
+            event
+              .with_time(this.currBasal.basalEndTime)
```

Comment:
Setting `time` without using `fillInUTCInfo` usually leads to unexpected side effects.